### PR TITLE
Do not fail completely on missing GStreamer

### PR DESCRIPTION
**Pull request by szpak**
If optional sound dependencies are not available, just print warning
and skip the sound part instead of failing completely.

It would prevent #47 and #52.

Btw, I am not a JavaScript developer. Feel free to fix naming convention or lame constructions.
Btw2, the warning in preferences could be "got prettier":

image

### CODE REVIEW

File: extension.js
Comment (szpak) [inline extension.js:129]: Extension UUID is not yet available.

File: prefs.js
Comment (szpak) [inline prefs.js:119]: I wasn't able to set the color just with properties.

**Comment by wbolster** [comment]
@BenignBeppe wdyt?

File: prefs.js
Comment (BenignBeppe) [inline prefs.js:135]: I don't think you need to log here too. There should be a message from extension.js.

File: prefs.js
Comment (BenignBeppe) [inline prefs.js:117]: This doesn't have any affect on the visibility, at least not on my machine. Instead you could add the label if playing is not supported. I think if you add the switch if it is supported and the warning message if not, it's even clearer to the user that sound can not be enabled. You could also put the label in the same position as the switch, but having it below also works.

File: prefs.js
Comment (BenignBeppe) [inline prefs.js:120]: Add `playingSoundNotSupportedLabel.set_line_wrap(true)` to wrap the text. Otherwise it forces the window to be wider than it needs to be.

File: extension.js
Comment (BenignBeppe) [inline extension.js:128]: Typo: GStremer -> GStreamer
Also in readme.

**Review by BenignBeppe — CHANGES_REQUESTED** [review]
This is definitely good to have. It would be even better if GStreamer could be added through dependencies, but I don't know if/how it possible with Gnome extensions.

File: prefs.js
Comment (szpak) [reply prefs.js]: Hmm, strange, It believe I checked it.

Ok, it seems that using `set_markup` "breaks" `visible`. Without that it works as expected. I switched to `set_visible()` and it works also with `set_markup()`.

Your proposal it definitely more user friendly, however, it is more complicated (with my zero-to-none skills in JavaScript :-) ). Eventually, it applied `set_sensitive()`, however, the state can be still "on":
image

Once the dependency is installed, it should just play the sounds (without switching it on manually).

File: prefs.js
Comment (szpak) [reply prefs.js:123]: I tried (with Gnome Shell 40.9) and it fails with:
> TypeError: playingSoundNotSupportedLabel.set_line_wrap is not a function

It seems to be available only in the Python bindings. In JS, I had to use `set_wrap(true)`.

**Comment by szpak** [comment]
> This is definitely good to have. It would be even better if GStreamer could be added through dependencies, but I don't know if/how it possible with Gnome extensions.

It could be applied for RPM/DEB with the extension, but I also haven't seen that mechanism for ZIP extensions downloaded from the "portal".

File: prefs.js
Comment (szpak) [reply prefs.js:138]: I added it as preferences can be shown also for disabled extensions. 

In fact, only the second log statement was important for me - I just don't like to hide exceptions (it is JavaScript - there could be something unexpected thrown ;-) ).

File: extension.js
Comment (szpak) [reply extension.js:15]: I have to think it over.

Comment (szpak) [reply extension.js:15]: I might miss some clever trick, but to prevent:
> ReferenceError: Gst is not defined

it would have to use multi-assignment on values returned from the method and I'm not sure if it would increase readability.

I also cannot move it after Extension as it would not be initialized.

Comment (BenignBeppe) [reply extension.js:15]: I did this:
```
const Extension = ExtensionUtils.getCurrentExtension();

let isPlayingSoundSupported = true;
let Gst;
let GstAudio
try {
    Gst = imports.gi.Gst;
    GstAudio = imports.gi.GstAudio;
} catch(e) {
    isPlayingSoundSupported = false;
    log(`${Extension.metadata.uuid}: Unable to import sound module. Playing sound is not available. Is GStremer package installed?`);
    log(`${Extension.metadata.uuid}: ${e}`);
}
```
Which I thought was a bit cleaner.

Comment (szpak) [reply extension.js:17]: I don''t know why I had `ReferenceError: Gst is not defined` before :-/. When I switched the order of statements again it works. I changed them.

However, personally, I prefer using constants instead of variables. The method name is self explaining. Nevertheless, if @wbolster prefers the construction you proposed, we can change it.

Comment (wbolster) [reply extension.js:17]: no strong preference: i like `const` but i also like `non-nullable` types 😉 so it's a trade-off. anyway, you can leave it as-is 👍🏼

File: extension.js
Comment (wbolster) [inline extension.js:128]: i like how this can help users fix their problems 👍🏼

**Review by wbolster — APPROVED** [review]

**Comment by wbolster** [comment]
lgtm after a cursory glance, but i'd like @BenignBeppe to have the final say since they looked before in more detail. please have a look again @BenignBeppe if you have time 🙏🏼

File: prefs.js
Comment (BenignBeppe) [reply prefs.js:123]: Looks like this is another case where we need to check Gnome shell version. I get:
`TypeError: playingSoundNotSupportedLabel.set_wrap is not a function`
on 3.36.9😞.

Comment (BenignBeppe) [reply prefs.js:123]: Maybe that's also why `set_visible()` doesn't do anything for me? The warning still shows, even if I explicitly do `set_visible(false)`.

File: extension.js
Comment (BenignBeppe) [reply extension.js:17]: This was more a personal preference so you can keep it as it is. I don't see any actual issues with it (it's just not how I would have done in🙂).

**Comment by BenignBeppe** [comment]
Looks good except there is still an issue with warning being shown when it shouldn't with Gnome shell 3.36.9 (and wrap function causes an error).

File: prefs.js
Comment (szpak) [reply prefs.js:123]: Could you paste that section in the way that works also with 3.36? Maybe we will be able to do some sensible conditional work?

Comment (BenignBeppe) [reply prefs.js:123]: Wrapping the whole label code (line 117 - 123) in `    if(!isPlayingSoundSupported)` solves visibility. Adding
```
if (GTK_VERSION == 3) {
  playingSoundNotSupportedLabel.set_line_wrap(true);
} else {
  playingSoundNotSupportedLabel.set_wrap(true);
}
```
solves the wrapping. Note again that this is with Gnome shell 3.36.9, so it needs to be tested with other versions.

Comment (BenignBeppe) [reply prefs.js:123]: The whole thing would look like this:
```
  if(!isPlayingSoundSupported) {
    const playingSoundNotSupportedLabel = new Gtk.Label({
      halign: Gtk.Align.START,
    });
    playingSoundNotSupportedLabel.set_markup("<span foreground='red'>WARNING. Playing sound is not supported on this system. Is GStreamer package installed?</span>");
    if (GTK_VERSION == 3) {
      playingSoundNotSupportedLabel.set_line_wrap(true);
    } else {
      playingSoundNotSupportedLabel.set_wrap(true);
    }
    prefsWidget.attach(playingSoundNotSupportedLabel, 0, 5, 1, 1);
  }
```

Comment (szpak) [reply prefs.js]: Cool, thanks. It works also with 40.9.

Comment (BenignBeppe) [reply prefs.js]: 👍

File: prefs.js
Comment (BenignBeppe) [inline prefs.js:143]: typo: GStremer -> GStreamer

Comment (szpak) [reply prefs.js:143]: Hehe, I fixed it already in the extensions.js :).

**Review by BenignBeppe — APPROVED** [review]

**Pull request merged**
